src/trio/testing/_raises_group.py

```diff
@@ -0,0 +1,305 @@
+else:
+    SuperClass = Generic
+
+
+@final
+class RaisesGroup(ContextManager[ExceptionInfo[BaseExceptionGroup[E]]], SuperClass[E]):
+    # needed for pyright, since BaseExceptionGroup.__new__ takes two arguments
+    if TYPE_CHECKING:
+
+        def __new__(cls, *args: object, **kwargs: object) -> RaisesGroup[E]:
+            ...
+
+    def __init__(
+        self,
+        exceptions: type[E] | Matcher[E] | E,
+        *args: type[E] | Matcher[E] | E,
+        strict: bool = True,
+        match: str | Pattern[str] | None = None,
+        check: Callable[[BaseExceptionGroup[E]], bool] | None = None,
+    ):
+        self.expected_exceptions: tuple[type[E] | Matcher[E] | E, ...] = (
+            exceptions,
+            *args,
+        )
+        self.strict = strict
+        self.match_expr = match
+        self.check = check
+        self.is_baseexceptiongroup = False
+
+        for exc in self.expected_exceptions:
+            if isinstance(exc, RaisesGroup):
+                if not strict:
+                    raise ValueError(
+                        "You cannot specify a nested structure inside a RaisesGroup with"
+                        " strict=False"
+                    )
+                self.is_baseexceptiongroup |= exc.is_baseexceptiongroup
+            elif isinstance(exc, Matcher):
+                if exc.exception_type is None:
+                    continue
+                # Matcher __init__ assures it's a subclass of BaseException
+                self.is_baseexceptiongroup |= not issubclass(
+                    exc.exception_type, Exception
+                )
+            elif isinstance(exc, type) and issubclass(exc, BaseException):
+                self.is_baseexceptiongroup |= not issubclass(exc, Exception)
+            else:
+                raise ValueError(
+                    "Invalid argument {exc} must be exception type, Matcher, or"
+                    " RaisesGroup."
+                )
+
+    def __enter__(self) -> ExceptionInfo[BaseExceptionGroup[E]]:
+        self.excinfo: ExceptionInfo[BaseExceptionGroup[E]] = ExceptionInfo.for_later()
+        return self.excinfo
+
+    def _unroll_exceptions(
+        self, exceptions: Iterable[BaseException]
+    ) -> Iterable[BaseException]:
+        """Used in non-strict mode."""
+        res: list[BaseException] = []
+        for exc in exceptions:
+            if isinstance(exc, BaseExceptionGroup):
+                res.extend(self._unroll_exceptions(exc.exceptions))
+
+            else:
+                res.append(exc)
+        return res
+
+    def matches(
+        self,
+        exc_val: BaseException | None,
+    ) -> TypeGuard[BaseExceptionGroup[E]]:
+        if exc_val is None:
+            return False
+        # TODO: print/raise why a match fails, in a way that works properly in nested cases
+        # maybe have a list of strings logging failed matches, that __exit__ can
+        # recursively step through and print on a failing match.
```

Comment:
Do we need to do this before merging?

Reply:
I'd say it depends on expected usage - if it's a helper that lots of people start using, and/or if it takes a long time until it's supplied by pytest, then it's probably worth chugging through now.

But I don't know how much of it would be applicable to any eventual solution in pytest, or if anybody cares to use this, so I'm hesitant to go all-in on polishing this.